clang/test/CIR/CodeGen/dynamic-cast.cpp

```diff
@@ -0,0 +1,52 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -std=c++17 -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+struct Base {
+  virtual ~Base();
+};
+// CHECK: !ty_22Base22 = !cir.struct
+
+struct Derived : Base {};
+// CHECK: !ty_22Derived22 = !cir.struct
+
+// CHECK: cir.func private @__dynamic_cast(!cir.ptr<!void>, !cir.ptr<!u8i>, !cir.ptr<!u8i>, !s64i) -> !cir.ptr<!void>
+
+Derived *ptr_cast(Base *b) {
+  return dynamic_cast<Derived *>(b);
+}
+//      CHECK: cir.func @_Z8ptr_castP4Base
+//      CHECK:   %[[#V1:]] = cir.load %{{.+}} : cir.ptr <!cir.ptr<!ty_22Base22>>, !cir.ptr<!ty_22Base22>
+// CHECK-NEXT:   %[[#V2:]] = cir.cast(ptr_to_bool, %[[#V1]] : !cir.ptr<!ty_22Base22>), !cir.bool
+// CHECK-NEXT:   %[[#V3:]] = cir.unary(not, %[[#V2]]) : !cir.bool, !cir.bool
+// CHECK-NEXT:   %{{.+}} = cir.ternary(%[[#V3]], true {
+// CHECK-NEXT:     %[[#V4:]] = cir.const(#cir.ptr<null> : !cir.ptr<!ty_22Derived22>) : !cir.ptr<!ty_22Derived22>
+// CHECK-NEXT:     cir.yield %[[#V4]] : !cir.ptr<!ty_22Derived22>
+// CHECK-NEXT:   }, false {
+// CHECK-NEXT:     %[[#V5:]] = cir.const(#cir.global_view<@_ZTI4Base> : !cir.ptr<!u8i>) : !cir.ptr<!u8i>
+// CHECK-NEXT:     %[[#V6:]] = cir.const(#cir.global_view<@_ZTI7Derived> : !cir.ptr<!u8i>) : !cir.ptr<!u8i>
+// CHECK-NEXT:     %[[#V7:]] = cir.const(#cir.int<0> : !s64i) : !s64i
+// CHECK-NEXT:     %[[#V8:]] = cir.cast(bitcast, %2 : !cir.ptr<!ty_22Base22>), !cir.ptr<!void>
+// CHECK-NEXT:     %[[#V9:]] = cir.call @__dynamic_cast(%[[#V8]], %[[#V5]], %[[#V6]], %[[#V7]]) : (!cir.ptr<!void>, !cir.ptr<!u8i>, !cir.ptr<!u8i>, !s64i) -> !cir.ptr<!void>
+// CHECK-NEXT:     %[[#V10:]] = cir.cast(bitcast, %[[#V9]] : !cir.ptr<!void>), !cir.ptr<!ty_22Derived22>
```

Comment:
I would prefer if dynamic cast is a CIR operation (such `cir.dynamic_cast` we could skip the casts and attach the `#cir.global_view` directly), we then lowered to exactly this CIR from the current testcase during the `LoweringPrepare` pass.

This PR is totally fine first approach though. Can you do incremental work and raise to a `cir.dynamic_cast` in another PR after this one lands?

Reply:
> I would prefer if dynamic cast is a CIR operation (such cir.dynamic_cast we could skip the casts and attach the #cir.global_view directly). We can then lower to this CIR from current testcase (during the LoweringPrepare pass).

Yes that is the longer plan. I'll do the incremental work to invent a new operation for `dynamic_cast` in future PRs :)